### PR TITLE
Use the BINARY flag instead of COLLATE *_bin

### LINKED ISSUE #1286: Allow %s_bin in SQL field definitions
<a href="https://github.com/fritzmg"></img></a> [Comment](https://github.com/contao/newsletter-bundle/issues/13#issuecomment-355225031) by @fritzmg
January 4th, 2018, 09:29 GMT

Btw. why do all the `alias` fields use a hardcoded `utf8mb4_bin` collation now anyway? Even if you set
```yml
doctrine:
    dbal:
        connections:
            default:
                default_table_options:
                    charset: utf8
                    collate: utf8_unicode_ci
```
the collation of these fields would still be `utf8mb4_bin`.

### PR DESCRIPTION
**Pull request by leofeyer**
This PR fixes #1286 as discussed with @aschempp. It handles the following cases:

* `'sql' => "varchar(128) BINARY NOT NULL default ''"`
* `'sql' => ['type' => 'string', 'length' => 128, 'not_null' => false, 'default' => '', 'customSchemaOptions' => ['case_sensitive' => true]]`

### CODE REVIEW

File: src/Doctrine/Schema/DcaSchemaProvider.php [hunk 1]
```diff
@@ -190,6 +206,11 @@ private function parseColumnSql(Table $table, string $columnName, string $sql):
             $collation = $match[1];
         }
 
+        // Use the binary collation if the BINARY flag is set (see #1286)
+        if (preg_match('/\bbinary\b/i', $def, $match)) {
```
Comment (aschempp) [inline src/Doctrine/Schema/DcaSchemaProvider.php]: Should this be supported for non-VARCHAR fields? Also, if a collation is already set, I assume it should not be overridden?

File: src/Doctrine/Schema/DcaSchemaProvider.php [hunk 2]
```diff
@@ -387,8 +408,8 @@ private function getSqlDefinitions(): array
     private function getMaximumIndexLength(Table $table, string $column): int
     {
         $indexLength = $this->getDefaultIndexLength($table);
+        $collation = $table->getOption('collate');
```
Comment (aschempp) [inline src/Doctrine/Schema/DcaSchemaProvider.php]: Are you sure this is correct? The documentation says this should be `collation` (http://docs.doctrine-project.org/projects/doctrine-dbal/en/latest/reference/schema-representation.html#vendor-specific-options). Not sure if that's for fields only though.

Comment (leofeyer) [reply src/Doctrine/Schema/DcaSchemaProvider.php]: Yes, I am sure. The default table options are merged:

File: src/Doctrine/Schema/DcaSchemaProvider.php (hunk 1 above)
Comment (leofeyer) [reply src/Doctrine/Schema/DcaSchemaProvider.php]: > Should this be supported for non-VARCHAR fields?

Yes, of course. It should be supported for every field that supports collations (usually all text fields).

> Also, if a collation is already set, I assume it should not be overridden?

I don't think so. Having both `COLLATE` and `BINARY` is a configuration error.

File: src/Doctrine/Schema/DcaSchemaProvider.php (hunk 2 above)
Comment (m-vo) [reply src/Doctrine/Schema/DcaSchemaProvider.php]: I'd stumbled across this as well - ``collation`` for fields ``colate`` for tables, extremely unintuitive... :-/

File: src/Doctrine/Schema/DcaSchemaProvider.php (hunk 1 above)
Comment (aschempp) [reply src/Doctrine/Schema/DcaSchemaProvider.php]: But then we should invert the if's, so that `COLLATE` overrides `BINARY` imho.

Comment (leofeyer) [reply src/Doctrine/Schema/DcaSchemaProvider.php]: No. If the `BINARY` flag is set, it means "use the binary collation". Doesn't matter what collation is given IMHO.

But why don't we just throw an exception? It is a configuration error.

Comment (leofeyer) [reply src/Doctrine/Schema/DcaSchemaProvider.php]: Changed in d8f4653d71baba848b2b4b598dd2dc7d5a341e14.

Comment (leofeyer) [reply src/Doctrine/Schema/DcaSchemaProvider.php]: Nope, we should not throw an error message but do the same as MySQL does if we execute the following query:

```sql
ALTER TABLE `tl_article` ADD `test` VARCHAR(255) BINARY COLLATE utf8mb4_unicode_ci NOT NULL;
```

This is how the field is created:



The `binary` flag has precedence over the `collate` function, so the initial code was correct.
